lib/services/watcher.go

```diff
@@ -1681,55 +1682,59 @@ type SAMLIdPServiceProviderWatcherConfig struct {
 
 // CheckAndSetDefaults checks parameters and sets default values.
 func (cfg *SAMLIdPServiceProviderWatcherConfig) CheckAndSetDefaults() error {
-	if err := cfg.ResourceWatcherConfig.CheckAndSetDefaults(); err != nil {
+	if err := cfg.RWCfg.CheckAndSetDefaults(); err != nil {
 		return trace.Wrap(err)
 	}
-	if cfg.SAMLIdPServiceProviders == nil {
-		getter, ok := cfg.Client.(SAMLIdPServiceProviders)
+	if cfg.spService == nil {
+		getter, ok := cfg.RWCfg.Client.(SAMLIdPServiceProviders)
 		if !ok {
 			return trace.BadParameter("missing parameter SAMLIdPServiceProviders and Client not usable as SAMLIdPServiceProviders")
 		}
-		cfg.SAMLIdPServiceProviders = getter
-	}
-	if cfg.PageSize == 0 {
-		cfg.PageSize = 200
+		cfg.spService = getter
 	}
 	if cfg.SAMLIdPServiceProvidersC == nil {
 		cfg.SAMLIdPServiceProvidersC = make(chan types.SAMLIdPServiceProviders)
 	}
 	return nil
 }
 
-// NewSAMLIdPServiceProviderWatcher returns a new instance of SAMLIdPServiceProviderWatcher.
+// NewSAMLIdPServiceProviderWatcher returns a new instance of SAMLIdPServiceProviderWatcher. The context here will be used to
+// exit early from the resource watcher if needed.
 func NewSAMLIdPServiceProviderWatcher(ctx context.Context, cfg SAMLIdPServiceProviderWatcherConfig) (*SAMLIdPServiceProviderWatcher, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 	collector := &samlIDPServiceProviderCollector{
-		SAMLIdPServiceProviderWatcherConfig: cfg,
-		initializationC:                     make(chan struct{}),
+		Log:             cfg.RWCfg.Log,
+		Cfg:             cfg,
+		initializationC: make(chan struct{}),
 	}
-	watcher, err := newResourceWatcher(ctx, collector, cfg.ResourceWatcherConfig)
+	watcher, err := newResourceWatcher(ctx, collector, cfg.RWCfg)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return &SAMLIdPServiceProviderWatcher{watcher, collector}, nil
+	return &SAMLIdPServiceProviderWatcher{
+		ResourceWatcher: watcher,
+		Collector:       collector,
+	}, nil
 }
 
 // SAMLIdPServiceProviderWatcher is built on top of resourceWatcher to monitor SAML IdP service provider resources.
 type SAMLIdPServiceProviderWatcher struct {
-	*resourceWatcher
-	*samlIDPServiceProviderCollector
+	ResourceWatcher *resourceWatcher
+	Collector       *samlIDPServiceProviderCollector
 }
 
 // samlIDPServiceProviderCollector accompanies resourceWatcher when monitoring SAML IdP service provider resources.
 type samlIDPServiceProviderCollector struct {
+	// Log is a logger.
+	Log logrus.FieldLogger
```

Comment:
Ditto for private fields, no reason for public here.

Reply:
These fields have been made private.